bug 957802: Document the rendering process

jwhitlock opened the pull request:
Document the steps of the rendering process, with a simple document for illustration. It's clear that the process evolved over the years, and I try to emphasize where the process has some unintended effects.

jwhitlock commented:
I looked at this again, tweaked some sentences, and I think it is ready for review.

I've had a mental model of the rendering process in my head for the last few years, it was good to get it written out and see where I was wrong about details. There are things I'd like to change, so I added a "Future Changes" section. This also feels inappropriate for documentation, so I'd support dropping it.

One of the reasons I took this on now was that @davidflanagan dropped the [rendering overview graphic](https://github.com/mdn/kumascript/blob/abf7a36761517f63b1a85fd87ae76407868c6de0/overview.png) in the KumaScript refactor. I didn't recreate it for this PR. If more flowcharts would help, I can create some more.

The other reason was a request by @wbamberg and @escattone  in [bug 1519582](https://bugzilla.mozilla.org/show_bug.cgi?id=1519582) to change the ``$api`` endpoint to return unbleached content. I'm against serving unbleached content. It appears that MDN's policy has been "store what the user wrote, but don't display unsafe content", and that there are several competing strategies to perform just-in-time cleaning of content. I think we should move toward "store what the user wrote, after some basic processing", which would unify several revision formats.

I think there are a few changes we'll make to rendering in the future, and we can update this document as we make them.

davidflanagan commented on docs/rendering.rst:
```diff
@@ -0,0 +1,783 @@
+``on``, such as the scripting attempt ``onclick``. However, none of these
+should remain in the *cleaned content*.
+
+Live sample
```
Can you mention how live samples are related to interactive examples and the codepen interactive examples? Is this an outdated thing, or are we still using it? It might be helpful to have a link to a document that includes one of these in addition to the link to the docs.

I also don't really get what stage of the rendering process these are handled in. What prevents the live samples <pre> sections from showing up in the body html? (And would this section be better if it was right after the body html and quick links html section?)

jwhitlock replied:
The [Live samples](https://developer.mozilla.org/en-US/docs/MDN/Contribute/Structures/Live_samples) page includes a [demo](https://developer.mozilla.org/en-US/docs/MDN/Contribute/Structures/Live_samples#Live_sample_demo), which I'll link to.  We're still using it, and you can look for ``EmbedLiveSample`` on the [macro dashboard](https://developer.mozilla.org/en-US/dashboards/macros) to find usage (1171 English pages).

The ``<pre>`` sections often show up in the body HTML. The goal is usually to say "Here's some CSS, and here's what it looks like".

I'm currently splitting the document between the "main line" rendering sections and the alternate rendering sections. Let's see if that still makes sense once the whole thing is published.

I'll add all this to the doc, along with a quick list of the other sample types.

davidflanagan commented on docs/rendering.rst:
```diff
@@ -0,0 +1,783 @@
+* Content from editing could be normalized and filtered before storing as the
+  *revision content*. This may unify the *re-edit format*, *diff format*, and
+  *cleaned content*
+* Add IDs immediately to the *revision content*, rather than wait for the
```
Nit: you switch out of subjunctive mode here.  Maybe start this bullet and the next with "We could add"

davidflanagan commented on docs/rendering.rst:
```diff
@@ -0,0 +1,783 @@
+* Add more consistent ways to access and generate content, rather than
+  repeating filter logic in different forms and views.
+
+History
```
It is great to have all this history written down!

davidflanagan commented on docs/rendering.rst:
```diff
@@ -0,0 +1,783 @@
+
+.. _`Live samples`: https://developer.mozilla.org/en-US/docs/MDN/Contribute/Structures/Live_samples
+
+Future Changes
```
I like having this section here.

jwhitlock commented on docs/rendering.rst:
```diff
@@ -0,0 +1,783 @@
+* A document's new or updated content is stored as *revision content*
+* The content is processed to create *cleaned content*, with macros and safe HTML
+* The content is passed to KumaScript to create *KumaScript content*, with rendered macros
+* The content is processed to create *rendered content*, and is only safe HTML
```
I think I was trying for a narrative table of contents, which is a bit weak without links. I think it is a little clearer with the image, which is in-lined in the rendered version. I think I'll rework this section.

jwhitlock commented on docs/rendering.rst:
```diff
@@ -0,0 +1,783 @@
+
+Revision content
+================
+:doc:`CKEditor </ckeditor>` is used to provide a visual HTML editor for MDN
```
[:doc:](http://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#role-doc) is the Sphinx syntax to to link to another document, in this case the [CKEditor doc page](https://kuma.readthedocs.io/en/latest/ckeditor.html). It looks weird in GitHub, and works in the rendered doc.

jwhitlock commented on docs/rendering.rst:
```diff
@@ -0,0 +1,783 @@
+source
+   User-entered content, usually via CKEditor and ``$edit`` view
+
+   PUT ``$api``
```
True. I think I'll link to [PUT API](https://developer.mozilla.org/en-US/docs/MDN/Contribute/Tools/PUT_API).

jwhitlock commented on docs/rendering.rst:
```diff
@@ -0,0 +1,783 @@
+The PUT ``$API`` can also be used to add new revisions. This API is for staff
+only at this time.
+
+.. _`/en-US/docs/Sandbox/simple`: https://developer.mozilla.org/en-US/docs/Sandbox/simple
```
Yes, I lost a sentence when I was moving content around. I'll re-add it.

jwhitlock commented on docs/rendering.rst:
```diff
@@ -0,0 +1,783 @@
+database
+   ``wiki_document.html`` for current revision, not stored for historical revisions
+code
+   ``Document.get_html`` (current revision, cached), ``Revision.content_cleaned`` (any revision, dynamically generated)
```
Good idea, I'll update code references to full dotted paths, include ``()`` on function calls.

jwhitlock commented on docs/rendering.rst:
```diff
@@ -0,0 +1,783 @@
+In the sample document, the ``{{CSSRef}}`` macro renders a skeleton version of
+the full sidebar. On pages like `Media queries`_, the sidebar grows to several
+kilobytes, to bring in links to related CSS pages.  The
+``{{HTMLElement('div')}}`` requires page data, which is gathered via a HTTP
```
I'm expanding into multiple paragraphs, with additional links to the data used in each macro.

jwhitlock commented on docs/rendering.rst:
```diff
@@ -0,0 +1,783 @@
+``<p></p>`` elements from the *rendered content*. This can cause annoying
+empty space at the top of a document.
+
+IDs are injected into header elements (such as ``id="A_simple_document"``),
```
Good find, this is content left from a previous version.

jwhitlock commented on docs/rendering.rst:
```diff
@@ -0,0 +1,783 @@
+The "middle" of a wiki document is populated by the *body HTML*.
+
+source
+   Extracted from *rendered content*
```
Yes, the Body HTML is derived from the Rendered content, with some expensive processing (remove the sidebar, annotate links to new documents). It follows a "cache on first access" model, which is kicked off when the Rendered content is available.

jwhitlock commented on docs/rendering.rst:
```diff
@@ -0,0 +1,783 @@
+The sidebar, on pages that include it, is populated from the *quick links html*.
+
+source
+   Extracted from *rendered content*
```
I'll add "cached in the database on first access"

jwhitlock commented on docs/rendering.rst:
```diff
@@ -0,0 +1,783 @@
+
+Raw content
+===========
+A ``?raw`` parameter can be added to the end of a document to request the
```
I've tried it both ways. I think someone reading top to bottom would be better served by reading the main flow first, then jumping around the alternate flows. I hope adding more inter-document links makes it easier to jump around.

jwhitlock commented:
Thanks for the detailed review @davidflanagan. I've re-written large sections of the content.

* Rewrite the "contents" section to link to the different sections and use the active voice
* Use links to the simple sample document ([/en-US/docs/Sandbox/simple](https://developer.mozilla.org/en-US/docs/Sandbox/simple)) as live demos of different views, rather than shortcuts like [$api](https://developer.mozilla.org/en-US/docs/Sandbox/simple$api). I feel silly for calling this example simple now.
* Use Sphinx ``:ref:`` links to refer to other content sections (screenshot below)
* Change "on MDN" to "displayed on MDN" in content summary def list
* Standardize to one item per line in content summary def list
* Use full dotted path to attributes and methods to make it easier to find them. I resisted linking to GitHub because the links would quickly become stale.

Highlighted changes to specific sections:

* "Revision content": improve introduction to the sample document
* "KumaScript content: Expand the narrative section on how macros get their data.
* "ToC HTML": Explain when the "Jump To" bar is omitted or hidden
* "Summary Text": Expand how the summary text is picked
* "Diff Format": rewrite, and  explain how unsafe content is escaped for display
* "Live sample": Explain the live sample server, how the iframe is inserted, how code sections are identified, and provide a short reference to other sample types
* "Future Changes": Keep subjunctive mode
* "History": CVS not CSV, rewrite DekiWiki and Content Zones

I also re-read [Ryan's Discourse Post](https://discourse.mozilla.org/t/automating-page-changes-with-the-put-api/29842/8), and it's clear I'm about 7 months behind 😢 . I renamed "Cleaned content" to "Bleached content", and "Re-edit content" to "Triaged content", to closer match the naming scheme from that post.

I added a section on "Search content", since that is another way the content gets sliced and diced. I'm mentally thinking what an offline rendering process would look like, and search could still be part of it.

Finally, I updated the diagrams, and switched from JPEG to PNG for the non-SVG version. This gets used when Read The Docs generates some formats (PDF?)

The ``:ref:`` links are not rendered by GitHub, you would need to [generate them locally](file:///Users/john/src/kuma/docs/_build/html/documentation.html) to see the links. When I run it locally, it warns about missing links, and it renders cleanly. They pick up the name of the section they are attached to. Here's what the contents section looks like, rendered:

![pr 5216 - rendered](https://user-images.githubusercontent.com/286017/52067812-021bf780-2541-11e9-876e-a3d94108dc2b.png)

@davidflanagan, please take another look. @escattone, I'd like you to read as well, but it may be easier to review after merging and publishing.

jwhitlock commented:
This is a big document, and a bit ugly in GitHub. I'd support merging so it will appear rendered on ReadTheDocs, and making further changes in a new PR. A "Squash and Merge" feels appropriate.
